state/protocol/badger/mutator.go

```diff
@@ -511,6 +517,110 @@ func (m *Mutator) sealExtend(candidate *flow.Block) (*flow.Seal, error) {
+
+	var sealedHeight uint64
+	err := m.state.db.View(operation.RetrieveSealedHeight(&sealedHeight))
+	if err != nil {
+		return fmt.Errorf("could no retrieve sealed height: %w", err)
+	}
+
+	// forkBlocks is used to keep the IDs of the blocks we iterate through. We
+	// use it to skip receipts that are not for blocks in the fork.
+	forkBlocks := make(map[flow.Identifier]struct{})
+
+	// Create a lookup table of all the receipts that are already in blocks.
+	// This will be used to filter out duplicates.
+	lookup := make(map[flow.Identifier]struct{})
+
+	// loop through the fork backwards, from parent to last sealed, and keep
+	// track of blocks and receipts visited on the way.
+	ancestorID := header.ParentID
+	for {
+
+		// keep track of blocks we iterate over
+		forkBlocks[ancestorID] = struct{}{}
+
+		ancestor, err := m.state.headers.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor header (%x): %w", ancestorID, err)
+		}
+
+		index, err := m.state.index.ByBlockID(ancestorID)
+		if err != nil {
+			return fmt.Errorf("could not retrieve ancestor index (%x): %w", ancestorID, err)
+		}
+
+		// keep track of all receipts we iterate over
+		for _, recID := range index.ReceiptIDs {
+			lookup[recID] = struct{}{}
+		}
+
+		// break out when we reach the sealed height
+		if ancestor.Height <= sealedHeight {
+			break
+		}
+
+		ancestorID = ancestor.ParentID
+	}
+
+	// prevReceiptHeight is used to check that receipts within a payload are
+	// sorted by block height. For each receipt in the payload, we check that
+	// its block height is greater or equal than the previous receipt.
+	var prevReceiptHeight uint64 = 0
+
+	// check each receipt included in the payload for duplication
+	for _, receipt := range payload.Receipts {
+
+		// if the receipt was already included before, error
+		_, duplicated := lookup[receipt.ID()]
+		if duplicated {
+			return state.NewInvalidExtensionErrorf("payload includes duplicate receipt (%x)", receipt.ID())
+		}
+
+		// if the receipt is not for a block on this fork, error
+		if _, ok := forkBlocks[receipt.ExecutionResult.BlockID]; !ok {
+			return state.NewInvalidExtensionErrorf("payload includes receipt for block not on fork (%x)", receipt.ExecutionResult.BlockID)
+		}
+
+		// get the reference block to check expiry
+		h, err := m.state.headers.ByBlockID(receipt.ExecutionResult.BlockID)
+		if err != nil {
+			return fmt.Errorf("could not get reference block (%x): %w", receipt.ExecutionResult.BlockID, err)
+		}
+
+		// check whether receipt is for block at or below the sealed and
+		// finalized height
+		if h.Height <= sealedHeight {
+			// Block has either already been sealed and finalized  _or_
+			// block is a _sibling_ of a sealed and finalized block.
+			// In either case, it's an error
+			return state.NewInvalidExtensionErrorf("payload includes receipt for block below sealed height (%x) (%d/%d)",
+				receipt.ExecutionResult.BlockID,
+				h.Height,
+				sealedHeight)
+		}
```

Comment:
This check is only necessary, because we include the block with `sealedHeight` in `forkBlocks`. Lets assume we exclude the block with `sealedHeight` as suggested in my comment above. Then: 
* per construction, `forkBlocks` only contains blocks whose height is _larger than_ `sealedHeight`
* i.e. any block which passes the check https://github.com/onflow/flow-go/blob/32812eafcd71f3fe44e8e95c8646f7ae2d74d50d/state/protocol/badger/mutator.go#L590-L593 must have height > `sealedHeight`